rust/src/tokio/lz4.rs

```diff
@@ -0,0 +1,132 @@
+use lz4::liblz4::{
+    check_error, LZ4FDecompressionContext, LZ4F_createDecompressionContext, LZ4F_decompress,
+    LZ4F_freeDecompressionContext, LZ4F_VERSION,
+};
+use std::io::{Error, ErrorKind, Result};
+use std::ptr;
+use tokio::io::{AsyncRead, ReadBuf};
+
+const BUFFER_SIZE: usize = 32 * 1024;
+
+#[derive(Debug)]
+struct DecoderContext {
+    c: LZ4FDecompressionContext,
+}
+
+// An equivalent of the lz4::Decoder `std::io::Read` wrapper for `tokio::io::AsyncRead`.
+// Code below is adapted from the https://github.com/bozaro/lz4-rs crate.
+#[derive(Debug)]
+pub struct Lz4Decoder<R> {
+    c: DecoderContext,
+    r: R,
+    buf: Box<[u8]>,
+    pos: usize,
+    len: usize,
+    next: usize,
+}
+
+impl<R: AsyncRead> Lz4Decoder<R> {
```

Comment:
at some point we can/should upstream this into the `async-compression` crate that we use for zstd. However, contributing to that crate involves digging through several layers of macro magic, and I don't have capacity to do that right now.